Fix max recursion bug by removing logging.log calls in `emit`

DylanRussell opened the pull request:
# Description

Remove logging.log calls from `BatchProcessor.emit`. Any log calls in that function can get routed back to `emit` and ultimately result in a maximum recursion depth exceeded exception.

# How Has This Been Tested?

Added a unit test to prevent this.

# Does This PR Require a Contrib Repo Change?

- [ ] Yes. - Link to PR: 
- [x ] No.

# Checklist:

- [ x] Followed the style guidelines of this project
- [ x] Changelogs have been updated
- [ x] Unit tests have been added
- [ x] Documentation has been updated

emdneto reviewed (approved):
Test is failing at main. LGTM with a changelog.

DylanRussell commented:
Added a changelog entry

tammy-baylis-swi reviewed (approved):
Thank you for this 🙂

aabmass commented on opentelemetry-sdk/tests/shared_internal/test_batch_processor.py:
```diff
@@ -211,3 +214,29 @@ def test_record_processor_is_garbage_collected(
+                }
+            ),
+        )
+        set_logger_provider(logger_provider)
```
is this line needed?

DylanRussell replied:
This isn't needed, removed

aabmass commented on opentelemetry-sdk/tests/shared_internal/test_batch_processor.py:
```diff
@@ -211,3 +214,29 @@ def test_record_processor_is_garbage_collected(
         # Then the reference to the processor should no longer exist
         assert weak_ref() is None
+
+    def test_logging_lib_not_invoked_in_emit(
```
nit this test is not really testing this specifically, it's just making sure there's no exception right?

It also runs for all of the types of batch processor, but I think it only applies to the logging one

DylanRussell replied:
Yes it's just testing there's no exception raised, which I verified does happen if logging.log is called from `emit`

Your right it's just the logging one.. moved this test to the BatchLogRecordProcessor only test suite..

aabmass commented on opentelemetry-sdk/tests/shared_internal/test_batch_processor.py:
```diff
@@ -211,3 +214,29 @@ def test_record_processor_is_garbage_collected(
+        handler = LoggingHandler(
+            level=logging.INFO, logger_provider=logger_provider
+        )
+        # Attach OTLP handler to root logger
+        logging.getLogger().addHandler(handler)
```
Don't think this is needed for the test

aabmass replied:
Ah nvm I see it's needed for the repro, but you could attach it to the SDK logger specifically instead of the root. Please just update the test to cleanup the global state it changes.

DylanRussell replied:
This is needed to get logging.log messages routed back through the processor

aabmass reviewed (commented):
I think you can make the test much simpler to just test a minimal repro of the reported issue

DylanRussell commented:
Ok updated test to attach handler to SDK logger instead of root, and remove it after test is done.

aabmass commented on opentelemetry-sdk/tests/logs/test_export.py:
```diff
@@ -342,6 +342,29 @@ def test_emit_call_log_record(self):
+        processor.emit(EMPTY_LOG)
+        processor.shutdown()
+        processor.emit(EMPTY_LOG)
+        logging.getLogger("opentelemetry.sdk").removeHandler(handler)
```
sorry to be nitpicky but this won't run if there is an exception and it could cause unrelated tests to fail as red herrings since it swallows log messages.

can you add a contextmanager or try/finally block?

emdneto reviewed (commented):
@aabmass, just a reminder that the current main branch is different from [release/v1.33.x-0.54bx](https://github.com/open-telemetry/opentelemetry-python/tree/release/v1.33.x-0.54bx/opentelemetry-sdk/src/opentelemetry/sdk) branch in terms of the structure of directories for the SDK. In the release branch there's no `_shared_internal` directory because this change was merged this week. So, probably would be better that in the backport PR we just remove the log line or maybe would be worth @DylanRussell open the PR against the release branch instead of main.

DylanRussell commented:
Moved this to https://github.com/open-telemetry/opentelemetry-python/pull/4588 .. Had this code in my `main` branch by mistake.. moved it to another branch, but couldn't edit this PR to move to that branch. So just opened up a new PR
